internal/overlord/servstate/handlers.go

```diff
@@ -68,13 +68,16 @@ const (
 type serviceState string
 
 const (
-	stateInitial     serviceState = "initial"
-	stateStarting    serviceState = "starting"
-	stateRunning     serviceState = "running"
-	stateTerminating serviceState = "terminating"
-	stateKilling     serviceState = "killing"
-	stateStopped     serviceState = "stopped"
-	stateBackoff     serviceState = "backoff"
+	stateInitial          serviceState = "initial"
+	stateStarting         serviceState = "starting"
+	stateRunning          serviceState = "running"
+	stateTerminating      serviceState = "terminating"
+	stateKilling          serviceState = "killing"
+	stateStopped          serviceState = "stopped"
+	stateBackoff          serviceState = "backoff"
+	stateExited           serviceState = "exited"
+	stateCheckTerminating serviceState = "terminating (after check failure)"
+	stateCheckKilling     serviceState = "killing (after check failure)"
```

Comment:
Yeah, I was definitely on the fence about these too, and wondered if there's a better way. I've removed them now, and replaced the state they track with a `restarting bool` field. This makes the state machine simpler, though it does mean we have an additional state variable which is only valid in a couple of states. So I think strictly speaking those two additional states made sense as actual states, as they had different transitions/flow in the state machine, but pragmatically speaking, this may be simpler.

Note that the state machine is an implementation detail, not a list of user states. So these are to track internal states, not something that's exposed to the user.

Regarding `exiting`, that one's less clear -- it tracks the difference between manually stopped and the service just exiting itself. I tried removing it locally, but then you need to track another piece of state for "was manually stopped", because in `checkFailed` you don't want a check failure to backoff and restart a service if it's been manually stopped (only if it exited). Manual stop means stay stopped.

Reply:
Oh, I've also added a Graphviz diagram of the state machine. Not very nicely laid out, but here it is:

![state diagram](https://raw.githubusercontent.com/canonical/pebble/40e45233591c49618f20d85d3d91d6c3aa460e6f/internal/overlord/servstate/state-diagram.svg)